extension/ackextension/inmemory.go

```diff
@@ -0,0 +1,111 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ackextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/ackextension"
+
+import (
+	"context"
+	"sync/atomic"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	"go.opentelemetry.io/collector/component"
+)
+
+// InMemoryAckExtension is the in-memory implementation of the AckExtension
+type InMemoryAckExtension struct {
+	partitionMap                  *lru.Cache[string, *ackPartition]
+	maxNumPendingAcksPerPartition uint64
+}
+
+func newInMemoryAckExtension(conf *Config) *InMemoryAckExtension {
+	cache, _ := lru.New[string, *ackPartition](int(conf.MaxNumPartition))
+	return &InMemoryAckExtension{
+		partitionMap:                  cache,
+		maxNumPendingAcksPerPartition: defaultMaxNumPendingAcksPerPartition,
+	}
+}
+
+type ackPartition struct {
+	id     atomic.Uint64
+	ackMap *lru.Cache[uint64, bool]
+}
+
+func newAckStatus(maxPendingAcks uint64) *ackPartition {
+	id := uint64(0)
+
+	cache, _ := lru.New[uint64, bool](int(maxPendingAcks))
+	cache.Add(id, false)
+
+	as := ackPartition{
+		ackMap: cache,
+	}
+	as.id.Store(id)
+	return &as
+}
+
+func (as *ackPartition) nextAck() uint64 {
+	id := as.id.Add(1)
+	as.ackMap.Add(id, false)
+	return id
+}
+
+func (as *ackPartition) ack(key uint64) {
+	if _, ok := as.ackMap.Get(key); ok {
+		as.ackMap.Add(key, true)
+	}
+}
+
+func (as *ackPartition) computeAcks(ackIDs []uint64) map[uint64]bool {
+	result := make(map[uint64]bool, len(ackIDs))
+	for _, val := range ackIDs {
+		if isAcked, ok := as.ackMap.Get(val); ok && isAcked {
+			result[val] = true
+			as.ackMap.Remove(val)
+		} else {
+			result[val] = false
+		}
+	}
+
+	return result
+}
+
+// Start of InMemoryAckExtension does nothing and returns nil
+func (i *InMemoryAckExtension) Start(_ context.Context, _ component.Host) error {
+	return nil
+}
+
+// Shutdown of InMemoryAckExtension does nothing and returns nil
+func (i *InMemoryAckExtension) Shutdown(_ context.Context) error {
+	return nil
+}
+
+// ProcessEvent marks the beginning of processing an event. It generates an ack ID for the associated partition ID.
+func (i *InMemoryAckExtension) ProcessEvent(partitionID string) (ackID uint64) {
+	if val, ok := i.partitionMap.Get(partitionID); ok {
+		return val.nextAck()
+	}
+
+	i.partitionMap.Add(partitionID, newAckStatus(i.maxNumPendingAcksPerPartition))
+	return 0
```

Comment:
I am not a huge fan of the way 0 is handled. The ack generation is essentially is separated out from `nextAck` and `newAckPartition`.

There is also a race condition here. two goroutines with the same partition ID coming in can overwrite another at line 88. 
- two goroutines enter. partition not exist in the map, both hit line 88.
- 1 routine adds a new ack partition and return 0
- another one overwrites the previous partition and returns 0 as well
- two events now have ack ID of 0

Can we do it like this to handle both cases?
```
func newAckPartition(maxPendingAcks uint64) *ackPartition {
  cache, _ := lru.New[uint64, bool](int(maxPendingAcks))
    return ackPartition{
		ackMap: cache,
	}
}

func (i *InMemoryAckExtension) ProcessEvent(partitionID string) (ackID uint64) {
        if val, ok := i.partitionMap.Get(partitionID); ok {
		return val.nextAck()
	}

        // multiple goroutine can enter this block and overwrite partition from each other
        // last write win and nextAck() synchronize ack generation
	ii.partitionMap.Add(partitionID, newAckPartition(i.maxNumPendingAcksPerPartition))
	val, _ := i.partitionMap.Get(partitionID)
	return val.nextAck()
}
```
It is a bit ugly but it handles the concurrency and ack id generation responsibilities better. The result is we won't return ID of 0. it will always start with 1 which is ok.

an alternative is to use `ContainsOrAdd` but i think it would have a worse performance since we will create a new partition for every event:
```
func (i *InMemoryAckExtension) ProcessEvent(partitionID string) (ackID uint64) {
        i.partitionMap.ContainsOrAdd(partitionID, newAckPartition(i.maxNumPendingAcksPerPartition))
	val, _ := i.partitionMap.Get(partitionID)
	return val.nextAck()
}
```

Reply:
Thanks for fixing it. But can you add a test for the race condition?
- spawn multiple goroutines
- all of them call `ProcessEvent` with the same partition ID
- ensure ACK IDs generated are synchronized

Reply:
done